chain/ethhashlookup/eth_transaction_hash_lookup.go

```diff
@@ -78,31 +129,43 @@ func (ei *EthTxHashLookup) GetHashFromCid(c cid.Cid) (ethtypes.EthHash, error) {
 }
 
 func (ei *EthTxHashLookup) DeleteEntriesOlderThan(days int) (int64, error) {
-	res, err := ei.db.Exec("DELETE FROM eth_tx_hashes WHERE insertion_time < datetime('now', ?);", "-"+strconv.Itoa(days)+" day")
-	if err != nil {
-		return 0, err
+	if ei.db == nil {
+		return 0, xerrors.New("db closed")
 	}
 
-	return res.RowsAffected()
-}
-
-func NewTransactionHashLookup(ctx context.Context, path string) (*EthTxHashLookup, error) {
-	db, _, err := sqlite.Open(path)
+	res, err := ei.stmtDeleteOlderThan.Exec("-" + strconv.Itoa(days) + " day")
 	if err != nil {
-		return nil, xerrors.Errorf("failed to setup eth transaction hash lookup db: %w", err)
-	}
-
-	if err := sqlite.InitDb(ctx, "eth transaction hash lookup", db, ddls, []sqlite.MigrationFunc{}); err != nil {
-		_ = db.Close()
-		return nil, xerrors.Errorf("failed to init eth transaction hash lookup db: %w", err)
+		return 0, err
 	}
-
-	return &EthTxHashLookup{db: db}, nil
+	return res.RowsAffected()
 }
 
-func (ei *EthTxHashLookup) Close() error {
+func (ei *EthTxHashLookup) Close() (err error) {
 	if ei.db == nil {
 		return nil
 	}
-	return ei.db.Close()
+	// set the db to nil first to ensure further queries or inserts fail early,
+	// then close preparedstatements, then close the db
+	db := ei.db
```

Comment:
That's why I'm doing the close in the `defer` - it gets closed after the statements are closed, and the `ei.db = nil` is used then as the signal for any future attempted queries or inserts with the new `== nil` checks in those functions. Since we don't _set_ `ei.db` anywhere other than the constructor, checking for nil in those methods is an atomic operation so should be safe.

The potential race here is in closing the statements while trying to use them. But, my judgement here is that since Close() is only used when completely shutting down a node, the liklihood of a race is very minimal if we get the `ei.db = nil` in early before closing the statements. If there is a race, it'd show up in itests where we have to do this shutdown.

An alternative is to not close the statements at all. It's good practice but not strictly necessary, especially since we currently have no code pathways involving a Close() that also don't shut the process down entirely. We also don't close statements in other dbs.

Reply:
And, introducing a lock for each access just to cleanly handle `Close()` is a bit unfortunate. Maybe not the worst thing but locks have cost.

Reply:
OK, nevermind, I just saw another panic so the timing is still off. I've removed the statement closes entirely now.